pep-9999.rst

```diff
@@ -0,0 +1,173 @@
+of hacks currently used in ``typing``.
+
+
+Performance:
+------------
+
+The ``typing`` module is one of the heaviest and slowest modules in
+the standard library even with all the optimizations made. Mainly this is
+because subscripted generic types (see PEP 484 for definition of terms
+used in this PEP) are class objects (see also [1]_). The three main ways how
+the performance can be improved with the help of the proposed special methods:
+
+- Creation of generic classes is slow since the ``GenericMeta.__new__`` is
+  very slow, we will not need it anymore.
+
+- Very long MROs for generic classes will be twice shorter, they are present
+  because we duplicate the ``collections.abc`` inheritance chain
+  in ``typing``.
+
+- Time of instantiation of generic classes will be improved
+  (this is minor however).
+
+
+Metaclass conflicts:
+--------------------
+
+All generic types are instances of ``GenericMeta``, so if a user uses
+a custom metaclass, then it is hard to make a corresponding class generic.
+This is particularly hard for library classes that a user doesn't control.
+A workaround is to always mix-in ``GenericMeta``::
+
+  class AdHocMeta(GenericMeta, LibraryMeta):
+      pass
+
+  class UserClass(LibraryBase, Generic[T], metaclass=AdHocMeta):
+      ...
+
+but this is not always practical or even possible. With the help of the
+proposed special attributes the ``GenericMeta`` metaclass will not be needed.
+
+
+Hacks and bugs that will be removed by this proposal:
+-----------------------------------------------------
+
+- ``_generic_new`` hack that exists since ``__init__`` is not called on
+  instances with a type differing form the type whose ``__new__`` was called,
+  ``C[int]().__class__ is C``.
+
+- ``_next_in_mro`` speed hack will be not necessary since subscription will
+  not create new classes.
+
+- Ugly ``sys._getframe`` hack, this one is particularly nasty, since it looks
+  like we can't remove it without changes outside ``typing``.
+
+- Currently generics do dangerous things with private ABC caches
+  to fix large memory consumption that grows at least as O(N\ :sup:`2`),
+  see [2]_. This point is also important because it was recently proposed to
+  re-implement ``ABCMeta`` in C.
+
+- Problems with sharing attributes between subscripted generics,
+  see [3]_. Current solution already uses ``__getattr__`` and ``__setattr__``,
+  but it is still incomplete, and solving this without the current proposal
+  will be hard and will need ``__getattribute__``.
+
+- ``_no_slots_copy`` hack, where we clean-up the class dictionary on every
+  subscription thus allowing generics with ``__slots__``.
+
+- General complexity of the ``typing`` module, the new proposal will not
+  only allow to remove the above mentioned hacks/bugs, but also simplify
+  the implementation, so that it will be easier to maintain.
+
+
+Specification
+=============
+
+Idea of ``__class_getitem__`` is very simple, it is an exact analog of
```

Comment:
"The idea of ``__class_getitem__`` is simple: it is ..."

Also, perhaps it's not that simple. The spec doesn't indicate how it's called and doesn't give an example of how one should implement it. (It is a class method?) Maybe you can give a complete example, e.g.
```py
class MyList:
    def __getitem__(self, index):
        "Called e.g. for `a = MyList(...); x = a[index]`"
        <return the indexed item from the instance>
    @classmethod
    def __class_getitem__(cls, subscript):
        "Called e.g. for `t = MyList[subscript]`"
        <return an object representing a subscripted type>
```

Reply:
``__class_getitem__`` is automatically a class method (like ``__init_subclass__``), so that ``@classmethod`` decorator is not needed here. I will add this example.